src/Interpreters/ClusterDiscovery.cpp

```diff
@@ -0,0 +1,276 @@
+#include <algorithm>
+#include <chrono>
+#include <unordered_set>
+
+#include <base/getFQDNOrHostName.h>
+#include <base/logger_useful.h>
+
+#include <Common/DNSResolver.h>
+#include <Common/StringUtils/StringUtils.h>
+#include <Common/ZooKeeper/Types.h>
+#include <Common/setThreadName.h>
+
+#include <Core/ServerUUID.h>
+
+#include <Interpreters/Cluster.h>
+#include <Interpreters/ClusterDiscovery.h>
+#include <Interpreters/Context.h>
+
+namespace DB
+{
+
+namespace
+{
+
+constexpr size_t MAX_QUEUE_SIZE = 16;
+constexpr UInt64 QUEUE_OP_TIMEOUT_MS = 1000;
+
+fs::path getReplicasListPath(const String & zk_root)
```

Comment:
There can be not only replicas, but simply shards/nodes.

```suggestion
fs::path getShardsListPath(const String & zk_root)
```

Reply:
Done